lib/RefBackend/RefBackend.cpp

```diff
@@ -131,18 +136,40 @@ class MungeCallingConventions
   void runOnOperation() override {
     auto module = getOperation();
     OpBuilder b(module.getBodyRegion());
-
-    auto consumeFuncReturnFunc = b.create<FuncOp>(
-        module.getLoc(), "refbackend_consume_func_return",
+    auto consumeFuncReturnInt64Func = b.create<FuncOp>(
+        module.getLoc(), "refbackend_consume_int64_func_return",
+        FunctionType::get(
+            module.getContext(),
+            UnrankedMemRefType::get(b.getI64Type(), /*memorySpace=*/0), {}),
+        b.getStringAttr("private"));
+    auto consumeFuncReturnFloat32Func = b.create<FuncOp>(
+        module.getLoc(), "refbackend_consume_float32_func_return",
         FunctionType::get(
             module.getContext(),
             UnrankedMemRefType::get(b.getF32Type(), /*memorySpace=*/0), {}),
         b.getStringAttr("private"));
-    addEmitCInterfaceAttr(consumeFuncReturnFunc);
+    addEmitCInterfaceAttr(consumeFuncReturnInt64Func);
+    addEmitCInterfaceAttr(consumeFuncReturnFloat32Func);
     for (auto func : module.getOps<FuncOp>()) {
-      if (func == consumeFuncReturnFunc)
+      if (func == consumeFuncReturnInt64Func || func == consumeFuncReturnFloat32Func)
         continue;
-      if (failed(mungeFunction(func, consumeFuncReturnFunc)))
+      bool hadError = false;
+      func.walk([&](ReturnOp op) {
+        if (auto returnType = op.getOperandTypes()[0].dyn_cast<MemRefType>()) {
```

Comment:
I think we need to call `mungeFunction` unconditionally here because `mungeFunction` also deals with arguments. It might be better to use a DenseMap to map return element type to different consumeFuncReturn func and pass the map to `mungeFunction`.

Reply:
drive-by: I would recommend changing mungeFunction to take a `DenseMap</*returnElementType*/Type, FuncOp>` instead of a single FuncOp. It already has to do error-handling w.r.t. the return element type, which you can replace by seeing if the return element type is in the map. Effectively, presence in the map would serve a similar function to the element type check in isReturnMemRefTypeValid.

Reply:
Like this?

Reply:
yep!